### PR TITLE
IDENTITY-6136: Add the functionality to retrieve list of users from giving a identity claim

### PR DESCRIPTION
**Pull request by darshanasbg**
### Proposed changes in this pull request

[List all changes you want to add here. If you fixed an issue, please
add a reference to that issue as well.]

-

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/listener/IdentityStoreEventListener.java [hunk 1]
```diff
@@ -283,6 +285,45 @@ public boolean doPreSetUserClaimValue(String userName,
...
+        if (log.isDebugEnabled()) {
+            log.debug("doPreGetUserList executed in the IdentityStoreEventListener for claim URI: " + claimUri +
+                    " and claim value: " + claimValue);
+        }
+
+        try {
+            List<String> userIds = identityDataStore.list(claimUri, claimValue, userStoreManager);
+
+            // If this is the primary domain, all the users will be retrieved since the primary domain is not appended
+            // to the user name it the IDN table. So we have to filter users belongs to primary in Java level.
```
Comment (omindu) [inline components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/listener/IdentityStoreEventListener.java]: to the user name '**in**' the IDN ....

File: components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/store/JDBCIdentityDataStore.java [hunk 2]
```diff
@@ -266,6 +273,62 @@ public void remove(String userName, UserStoreManager userStoreManager) throws Id
...
 
+    @Override
+    public List<String> list(String claimUri, String claimValue,
+                             org.wso2.carbon.user.core.UserStoreManager userStoreManager) throws IdentityException {
+
+        List<String> userNames = new ArrayList<>();
+
+        // This is to support LDAP like queries, so this will provide support for only leading or trailing '*'
+        // filters. if the query has multiple '*' s the filter will ignore all.
+        if (claimValue.contains(QUERY_FILTER_STRING_ANY)) {
```
Comment (omindu) [inline components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/store/JDBCIdentityDataStore.java]: Can't we support `*foo*` type of search filter ? This is already supported when searching by other claims.

File: components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/store/JDBCIdentityDataStore.java [hunk 3]
```diff
@@ -266,6 +273,62 @@ public void remove(String userName, UserStoreManager userStoreManager) throws Id
...
+                // If the user has a domain, domain name is appended to the user name in the user name column. If
+                // we need to select user names only for current domain, we have to do a SQL like.
+                String userNameWithDomain;
+                if (StringUtils.equalsIgnoreCase(userStoreDomain, UserCoreConstants.PRIMARY_DEFAULT_DOMAIN_NAME)) {
+                    userNameWithDomain = SQL_FILTER_STRING_ANY;
+                } else {
+                    userNameWithDomain = userStoreDomain + UserCoreConstants.DOMAIN_SEPARATOR + SQL_FILTER_STRING_ANY;
+                }
+
+                preparedStatement.setString(4, userNameWithDomain);
```
Comment (omindu) [inline components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/store/JDBCIdentityDataStore.java]: Shall we add a debug log indicating the final username being searched ?

File: components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/store/JDBCIdentityDataStore.java [hunk 4]
```diff
@@ -266,6 +273,62 @@ public void remove(String userName, UserStoreManager userStoreManager) throws Id
...
+
+        // This is to support LDAP like queries, so this will provide support for only leading or trailing '*'
+        // filters. if the query has multiple '*' s the filter will ignore all.
+        if (claimValue.contains(QUERY_FILTER_STRING_ANY)) {
+            if ((claimValue.startsWith(QUERY_FILTER_STRING_ANY) &&
+                    !claimValue.substring(1).contains(QUERY_FILTER_STRING_ANY)) ||
+                    claimValue.endsWith(QUERY_FILTER_STRING_ANY) &&
+                            !claimValue.substring(0, claimValue.length() - 1).contains(QUERY_FILTER_STRING_ANY) &&
+                            claimValue.charAt(claimValue.length() - 2) != SQL_FILTER_CHAR_ESCAPE) {
+                claimValue = claimValue.replace(QUERY_FILTER_STRING_ANY, SQL_FILTER_STRING_ANY);
```
Comment (omindu) [inline components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/store/JDBCIdentityDataStore.java]: Shall we add a debug log indicating which final claimValue being searched ?

File: components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/store/JDBCIdentityDataStore.java [hunk 5]
```diff
@@ -266,6 +273,62 @@ public void remove(String userName, UserStoreManager userStoreManager) throws Id
...
+        List<String> userNames = new ArrayList<>();
+
+        // This is to support LDAP like queries, so this will provide support for only leading or trailing '*'
+        // filters. if the query has multiple '*' s the filter will ignore all.
+        if (claimValue.contains(QUERY_FILTER_STRING_ANY)) {
+            if ((claimValue.startsWith(QUERY_FILTER_STRING_ANY) &&
+                    !claimValue.substring(1).contains(QUERY_FILTER_STRING_ANY)) ||
+                    claimValue.endsWith(QUERY_FILTER_STRING_ANY) &&
+                            !claimValue.substring(0, claimValue.length() - 1).contains(QUERY_FILTER_STRING_ANY) &&
+                            claimValue.charAt(claimValue.length() - 2) != SQL_FILTER_CHAR_ESCAPE) {
```
Comment (omindu) [inline components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/store/JDBCIdentityDataStore.java]: Need some clarification regarding `claimValue.charAt(claimValue.length() - 2) != SQL_FILTER_CHAR_ESCAPE`

File: components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/store/JDBCIdentityDataStore.java (hunk 2 above)
Comment (JKAUSHALYA) [reply components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/store/JDBCIdentityDataStore.java]: Yes this is supported in this logic.

File: components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/store/JDBCIdentityDataStore.java (hunk 5 above)
Comment (JKAUSHALYA) [reply components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/store/JDBCIdentityDataStore.java]: Checking whether we have a escaped query filter character. "\*"

File: components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/store/JDBCIdentityDataStore.java (hunk 2 above)
Comment (darshanasbg) [reply components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/store/JDBCIdentityDataStore.java]: No its not supported in the current logic. Even the JDBC userstore manager does not support this. We have to fix both places if we are to support it.

Comment (darshanasbg) [reply components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/store/JDBCIdentityDataStore.java]: Created https://wso2.org/jira/browse/IDENTITY-6467 to track this seperately

File: components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/store/JDBCIdentityDataStore.java (hunk 4 above)
Comment (darshanasbg) [reply components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/store/JDBCIdentityDataStore.java]: Added all at once
https://github.com/wso2-extensions/identity-governance/pull/170/files#diff-eb8649c3eedd56d1e6d03a87058c6574R320

File: components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/store/JDBCIdentityDataStore.java (hunk 3 above)
Comment (darshanasbg) [reply components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/store/JDBCIdentityDataStore.java]: Done
https://github.com/wso2-extensions/identity-governance/pull/170/files#diff-eb8649c3eedd56d1e6d03a87058c6574R320

**Pull request merged**